src/pcp/gadgets.rs

```diff
@@ -192,6 +192,176 @@ impl<F: FieldElement> Gadget<F> for PolyEval<F> {
     }
 }
 
+/// The gadget for the MeanVarUnsignedVector type. It is not designed for general use.
+///
+/// MeanVarUnsigned is parameterized by a positive integer `bits`. Its arity is `2*bits+1`:
+///
+///  * The first `bits` values are interpreted as a vector `r_vec`.
+///  * The next `bits` values are interpreted as a vector `x_vec`.
+///  * The last value is interpreted as a singleton `x`.
+///
+/// The gadget is designed to output `x^2` *as long as* `x_vec` is a vector comprised of 0s and 1s.
+/// It first computes `w[l] = p(x_vec[l]) * r_vec[l]` for each `0 <= l < bits`, where `p(x) =
+/// x(x-1)`. It then computes `w = w[0] + ... + w[bits-1]` and returns `w + x^2`.
+///
+/// If `vec_x[l] == 1` or `vec_x[l] == 0`, then `p(vec_x[l]) == 0`; otherwise, `p(vec_x[l]) != 0`.
+/// The validity circuit for MeanVarUnsignedVector sets `r_vec[l] = r^(l+1)`, where `r` is a
+/// uniform random field element. This is ensures that, if `p(vec_x[l]) != 0` for some `l`, then
+/// `w != 0` with high probability.
+pub struct MeanVarUnsigned<F: FieldElement> {
+    poly: [F; 3],
+    /// Size of buffer for FFT operations.
+    n: usize,
+    /// Inverse of `n` in `F`.
+    n_inv: F,
+    /// The parameter that determines the circuit's arity.
+    bits: usize,
+}
+
+impl<F: FieldElement> MeanVarUnsigned<F> {
+    /// Constructs a MeanVarUnsigned gadget with parameter `bits`.
+    pub fn new(bits: usize, in_len: usize) -> Self {
+        let poly: Vec<F> = poly_range_check(0, 2);
+        let n = (in_len * 3).next_power_of_two();
+        let n_inv = F::from(F::Integer::try_from(n).unwrap()).inv();
+
+        Self {
+            poly: [poly[0], poly[1], poly[2]],
+            n,
+            n_inv,
+            bits,
+        }
+    }
+
+    pub(crate) fn call_poly_direct<V: AsRef<[F]>>(
+        &mut self,
+        outp: &mut [F],
+        inp: &[V],
+    ) -> Result<(), PcpError> {
+        let bits = self.bits;
+        let r_vec = &inp[..bits];
+        let x_vec = &inp[bits..2 * bits];
+        let x = inp[2 * bits].as_ref();
+
+        let z = poly_mul(x, x);
+        for i in 0..z.len() {
+            outp[i] = z[i];
+        }
+        for i in z.len()..outp.len() {
+            outp[i] = F::zero();
+        }
+
+        for l in 0..bits {
+            let mut z = r_vec[l].as_ref().to_vec();
+            for i in 0..3 {
+                for j in 0..z.len() {
+                    outp[j] += self.poly[i] * z[j];
+                }
+
+                if i < 2 {
+                    z = poly_mul(&z, x_vec[l].as_ref());
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    pub(crate) fn call_poly_fft<V: AsRef<[F]>>(
+        &mut self,
+        outp: &mut [F],
+        inp: &[V],
+    ) -> Result<(), PcpError> {
+        let bits = self.bits;
+        let n = self.n;
+        let r_vec = &inp[..bits];
+        let x_vec = &inp[bits..2 * bits];
+        let x = inp[2 * bits].as_ref();
```

Comment:
As in `call_poly_direct`, this should check that `inp.len() == 2 * bits + 1` and that `outp.len >= z.len`

Reply:
Here again, this check is performed by the caller: see https://github.com/abetterinternet/libprio-rs/pull/48/files#diff-e01a59bccdbfaacea57d3cf97d4ad2fc0b9d8f1d333aeaa0098b150ad60a7a02R352.